### ISSUE TITLE
Implement traits on Mint types instead of cgmath/nalgebra?

### ISSUE THREAD

**Issue by LPGhatguy** [olson-sean-k/plexus#21]
Kvark made a really cool crate called [mint](https://github.com/kvark/mint) that's designed to help ecosystem authors deal with math types generically without having to make specific implementations for each crate.

I didn't look too closely at Plexus, but it looks like it might be able to benefit from going through mint instead of enabling specific features for cgmath and nalgebra!

**Comment by olson-sean-k** [comment]
Thanks for opening an issue! I'm not sure I understand the intended usage of Mint, but it seems to me that it's meant to provide simple types for which other crates provide conversions. These types don't provide operations, just data, and can act as "glue" when crates disagree about which libraries to use for this type of math.

If that's true (?), then I'm not sure that Plexus should implement geometric traits for Mint types. One thing that stands out to me is that this would involve implementing operations that Mint explicitly doesn't support. For cgmath and nalgebra, the geometric trait implementations simply delegate to their respective crates.

However, I think it could be useful to implement conversions to/from Mint types for the default types that Plexus uses for mesh data (e.g., `Duplet` and `Triplet`). That would make it easier to use Mint types if users do not want to use cgmath or nalgebra types with Plexus. For example:

```rust
use mint::Point3;
use plexus::graph::MeshGraph;
use plexus::prelude::*;
use plexus::primitive::cube::Cube;

let mut graph = Cube::new()
    .polygons_with_position()
    .collect::<MeshGraph<Triplet<f32>>>();
let key = graph.vertices().nth(0).unwrap().key();
let position: Point3<f32> = graph.vertex(key).unwrap().geometry.into();
// `position` can be consumed by other crates or converted as necessary.
```

Thoughts?

**Comment by LPGhatguy** [comment]
I think that approach makes a lot of sense to me!

Another fairly significant advantage of implementing mint types is that consumers of Plexus can use a different version of cgmath or nalgebra than Plexus without having to do tricky conversions.

Plexus might be able to hide which math library it uses under the hood (nalgebra or cgmath) and just expose mint types. Users of every math library might need to write more explicit conversions though, which is sort of non-ideal.

**Comment by olson-sean-k** [comment]
Cool, I'll put together a change that implements conversions to/from Mint types for the types provided by Plexus.

> Plexus might be able to hide which math library it uses under the hood...

Plexus doesn't use or require any math crates under the hood. Everything is done in terms of the geometric traits that are part of its public API. Plexus provides its own minimal types that implement the geometric traits as well as conversions. This is actually a bit similar to Mint. Any dependencies on math crates or conversions are optional and controlled by crate features.

A lot of the documentation and examples use nalgebra (with implicit conversions), but that's only because I expect users to choose a suitable math crate for their own purposes instead of relying solely on anemic Plexus types. And I like nalgebra. :-)

> Another fairly significant advantage of implementing mint types is that consumers of Plexus can use a different version of cgmath or nalgebra than Plexus without having to do tricky conversions.

I think the same versioning issues also apply to Mint here, no?

My primary concern with implementing geometric traits for Mint types is the crate's intended usage. As I said above, it seems to me that providing outboard mathematical implementations and using Mint types internally is not the intention, but I don't know. @kvark?

**Comment by kvark** [comment]
Wow, you've got some interesting library here!

@olson-sean-k you are correct that Mint's purpose is to be the interchange format, it's explicitly avoiding any computational logic (as opposed to conversional, but only where it's moving bytes around, e.g. column major to row major matrix).

> I think the same versioning issues also apply to Mint here, no?

The idea is that `mint` is more stable than either of the specific libraries that convert to it, and we it isn't  - there is way more room for [semver trick](https://github.com/dtolnay/semver-trick) than with cgmath/euclid, because there isn't much logic to check for backwards compatibility. In fact, we used the trick when releasing v0.5 and nobody noticed :)

> providing outboard mathematical implementations and using Mint types internally is not the intention

There is nothing *technically* preventing a library to emerge that implements some math traits on `mint` types. In fact, I'd be happy to use it ;)

**Comment by olson-sean-k** [comment]
> The idea is that `mint` is more stable than either of the specific libraries that convert to it ... There is nothing _technically_ preventing a library to emerge that implements some math traits on `mint` types...

Thanks for taking a look, @kvark. It seems to me then that implementing some geometric traits should be harmless.

I've gone ahead and landed 75172b0, which implements conversions and geometric traits for Mint types. This is enabled by the `geometry-mint` feature. @LPGhatguy, feel free to close if you think that addresses the issue! If not, please let me know what other thoughts you have.

**Comment by olson-sean-k** [comment]
Closing this issue for now.
